File: packages/plugin-demo-python-http/backend.py
#!/usr/bin/env python3
"""
CodeExpander 插件后端 - http 模式 (Python)

协议：从环境变量 CODEEXPANDER_BACKEND_PORT 获取端口
     在 127.0.0.1:PORT 启动 HTTP 服务
     接收 POST / 请求，body 为 { method, payload }
     返回 JSON 响应

进程常驻，空闲约 5 分钟后退出
"""
import json
import os
from http.server import HTTPServer, BaseHTTPRequestHandler


def handle_request(method: str, payload: dict) -> dict:
    if method in ("run", "transform"):
        text = payload.get("text", "")
        action = payload.get("action", "upper")
        if action == "upper":
            result = text.upper()
        elif action == "lower":
            result = text.lower()
        elif action == "reverse":
            result = text[::-1]
        else:
            result = text
        return {"ok": True, "result": result, "action": action}
    if method == "ping":
        return {"ok": True, "pong": True, "mode": "http"}
    return {"ok": False, "error": f"Unknown method: {method}"}


class Handler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/":
            self.send_response(404)
            self.end_headers()
            return
        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length).decode("utf-8")
        try:
            data = json.loads(body)
            method = data.get("method", "run")
            payload = data.get("payload", {})
            response = handle_request(method, payload)
            out = json.dumps(response).encode("utf-8")
        except Exception as e:
            out = json.dumps({"ok": False, "error": str(e)}).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", len(out))
        self.end_headers()
        self.wfile.write(out)

    def log_message(self, *args):
        pass


def main():
    port = os.environ.get("CODEEXPANDER_BACKEND_PORT")
    if not port:
        print("CODEEXPANDER_BACKEND_PORT not set", file=__import__("sys").stderr)
        exit(1)
    server = HTTPServer(("127.0.0.1", int(port)), Handler)
    server.serve_forever()


if __name__ == "__main__":
    main()

File: packages/cli/templates/go-http/backend.go
// CodeExpander 插件后端 - http 模式 (Go)
// 从 CODEEXPANDER_BACKEND_PORT 获取端口，启动 HTTP 服务
// 进程常驻，多次调用复用同一进程
package main

import (
	"encoding/json"
	"net/http"
	"os"
	"strings"
)

type request struct {
	Method  string                 `json:"method"`
	Payload map[string]interface{} `json:"payload"`
}

type response struct {
	OK     bool   `json:"ok"`
	Error  string `json:"error,omitempty"`
	Result string `json:"result,omitempty"`
	Action string `json:"action,omitempty"`
	Pong   bool   `json:"pong,omitempty"`
	Mode   string `json:"mode,omitempty"`
}

func handleRequest(method string, payload map[string]interface{}) response {
	switch method {
	case "run", "transform":
		text, _ := payload["text"].(string)
		action, _ := payload["action"].(string)
		if action == "" {
			action = "upper"
		}
		var result string
		switch action {
		case "upper":
			result = strings.ToUpper(text)
		case "lower":
			result = strings.ToLower(text)
		case "reverse":
			runes := []rune(text)
			for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
				runes[i], runes[j] = runes[j], runes[i]
			}
			result = string(runes)
		default:
			result = text
		}
		return response{OK: true, Result: result, Action: action}
	case "ping":
		return response{OK: true, Pong: true, Mode: "http"}
	default:
		return response{OK: false, Error: "Unknown method: " + method}
	}
}

func main() {
	port := os.Getenv("CODEEXPANDER_BACKEND_PORT")
	if port == "" {
		os.Exit(1)
	}

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/" {
			http.NotFound(w, r)
			return
		}
		var req request
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			w.WriteHeader(http.StatusInternalServerError)
			json.NewEncoder(w).Encode(response{OK: false, Error: err.Error()})
			return
		}
		if req.Method == "" {
			req.Method = "run"
		}
		if req.Payload == nil {
			req.Payload = make(map[string]interface{})
		}
		res := handleRequest(req.Method, req.Payload)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(res)
	})

	http.ListenAndServe("127.0.0.1:"+port, nil)
}
